docs/rfcs/00-relationship-properties/relationship-properties.md

```diff
@@ -0,0 +1,341 @@
+
+## Usage example
+
+### Type definitions
+
+`@relationship` directive in each node on either side of the relationship has an additional `properties` argument containing the name of an interface which defines relationship properties.
+
+```graphql
+type Movie {
+    title: String!
+    actors: [Actor!]!
+        @relationship(type: "ACTED_IN", properties: "ActedIn", direction: IN)
+}
+
+type Actor {
+    name: String!
+    movies: [Movie!]!
+        @relationship(type: "ACTED_IN", properties: "ActedIn", direction: OUT)
+}
+
+interface ActedIn {
+    screenTime: Int!
+}
+```
+
+#### Notes
+
+-   Much like node types, property interfaces can be named however users see fit. If the same relationship type can be used for different node combinations, they can name them differently to distinguish between them.
+-   As per the current implementation, all types will be assumed to be node types and will have the relevant queries and mutations generated for them.
+-   An error will be thrown if a `@relationship` directive "pair" don't share the same `properties` value.
+
+### Schema changes
+
+#### New types for relationship properties and relationships
+
+Relationship types are automatically generated, one for each direction of a relationship. They are synonymous with the Relay concept of edges, with a field `node` which is the type at the other end of the relationship. The remaining fields will be from the properties interface which was specified in the `@relationship` directive, and the relationship type will implement that interface.
+
+```graphql
+interface ActedIn {
+    screenTime: Int!
+}
+
+type ActorMovieRelationship implements ActedIn {
+    node: Movie!
+    screenTime: Int!
+}
+
+type MovieActorRelationship implements ActedIn {
+    node: Actor!
+    screenTime: Int!
+}
+```
+
+#### New Connection types
+
+Connection types will be automatically generated for all relationships to facilitate relationship properties, and at a later time, cursor-based pagination.
+
+```graphql
+type ActorMoviesConnection {
+    edges: [ActorMovieRelationship!]!
+    pageInfo: PageInfo!
+}
+
+type MovieActorsConnection {
+    edges: [MovieActorRelationship!]!
+    pageInfo: PageInfo!
+}
+```
+
+#### New Where and Sort types
+
+There will be two new Where input types and two new Sort input types, one of each for relationship properties, and one of each to compose node and relationship properties together.
+
+For relationship properties:
+
+```graphql
+input ActedInWhere {
+    screenTime: Int
+    screenTime_NOT: Int
+    screenTime_LT: Int
+    screenTime_LTE: Int
+    screenTime_GT: Int
+    screenTime_GTE: Int
+    screenTime_IN: [Int]
+    screenTime_NOT_IN: [Int]
+    AND: [ActedInWhere!]
+    OR: [ActedInWhere!]
+}
+
+input ActedInSort {
+    screenTime: SortDirection
+}
+```
+
+And for composites of both node and relationship properties:
+
+```graphql
+input ActorMovieWhere {
```

Comment:
How would the auto-generated input names work if there are multiple distinct relationships between the same types, but each has a different set of properties?

For example:

```graphql
type Person {
    name: String!
    friends: [Person!]!
        @relationship(type: "HAS_FRIEND", properties: "HasFriend", direction: IN)
    enemies: [Person!]!
        @relationship(type: "HAS_ENEMY", properties: "HasEnemy", direction: IN)
}

interface HasFriend {
    since: DateTime!
    secretHandshake: String
}

interface HasEnemy {
    since: DateTime!
}
```

In this case, it's not clear if `PersonPersonWhere.relationship` should be `HasFriend` or `HasEnemy`.

Reply:
Very small change made to use the field name in that type instead of the type name, so in this case there would be `PersonFriendsWhere` and `PersonEnemiesWhere` containing the same "node where" but different "relationship where".

Reply:
I agree that this probably won't be an issue for the vast majority of users, and I think your updated approach is a reasonable default.

That said, since this package is generating schema to be merged with user-defined schema, I'd take it as a given that naming collisions will come up (whether caused by this issue or another).

For example:

```graphql
type User {
    name: String!
    status: UserStatus!
        @relationship(type: "HAS_STATUS", properties: "HasStatus", direction: IN)
}

interface HasStatus {
    since: DateTime!
}

type UserStatus {
    online: Boolean!
    message: String
}
```

This will cause a collision on the `UserStatusWhere` type.

I'm not sure it's possible to eliminate all possible collisions, so I'd suggest simply implementing a consistent and predictable way of handling them.

Reply:
Seems you're one step ahead of me every step of the way here! 😉 

These are so edge-casey, but you're absolutely right that we should be doing our absolute best to deconflict well in our solution.

Perhaps we need to be classifying `Where` input types by what they apply to, so in the above example:

- `UserWhere` and `UserStatusWhere` - no need to break these for everyone
- `HasStatusRelationshipWhere` - for matching on relationship properties
- `UserStatusConnectionWhere` - composing the above two together

Yes, slightly longer input type names, but should massively reduce naming collisions.

Reply:
> Perhaps we need to be classifying Where input types by what they apply to

Yes, I think that makes sense.

I might go with just `HasStatusWhere` instead of `HasStatusRelationshipWhere` though.

Reply:
Nevermind, I see why that doesn't work

Reply:
Actually, it does. Sorry.

With the change to `UserStatusConnectionWhere`, we should be able to do `HasStatusWhere` instead of `HasStatusRelationshipWhere`

Reply:
Ahaha, it's a bit of a mindwarp right, but you're right, changing to `*ConnectionWhere` automatically deconflicts anything which would have clashed with `*RelationshipWhere`.

Did just double check my sanity to see if you can name an interface the same as a type and indeed you can't, so `HasStatusWhere` and `UserStatusConnectionWhere` it is!

Thanks for working this through with me, will update the RFC either over the weekend or on Monday. 🙂